File: web/src/types/unifiedTestCase.ts
// 统一测试用例相关类型定义

import {
  PriorityLevel,
  TaskStatus,
  PaginationConfig,
  PaginationResponse,
  BaseApiResponse,
  SortDirection,
  SortField
} from './common';

// 直接定义统一的枚举以避免导入冲突
export const UnifiedTestCaseStatus = {
  DRAFT: 'draft',
  APPROVED: 'approved',
  REJECTED: 'rejected',
  COMPLETED: 'completed'  // 添加后端COMPLETED状态
} as const;

export const UnifiedTestCaseStage = {
  TEST_POINT: 'test_point',  // 改为小写单数，与后端一致
  TEST_CASE: 'test_case'       // 改为小写单数，与后端一致
} as const;

// 类型定义
export type UnifiedTestCaseStatus = 'draft' | 'approved' | 'rejected' | 'completed';
export type UnifiedTestCaseStage = 'test_point' | 'test_case';

// 类型别名
export type TestCaseStatus = UnifiedTestCaseStatus; // 向后兼容别名

export interface UnifiedTestCaseBase {
  name: string;
  description?: string;
  priority: PriorityLevel;
  status?: UnifiedTestCaseStatus; // 改为可选，支持默认值
}

export interface UnifiedTestCaseCreate extends UnifiedTestCaseBase {
  project_id: number;
  business_type: string;
  test_case_id: string; // 与后端 Pydantic alias 匹配
  module?: string;
  functional_module?: string;
  functional_domain?: string;
  preconditions?: string;  // 发送给后端时需要JSON字符串格式
  steps?: Array<{
    step_number: number;
    action: string;
    expected?: string;
  }>;
  expected_result?: string;
  remarks?: string;
  entity_order?: number;
}

export interface UnifiedTestCaseUpdate {
  name?: string;
  description?: string;
  business_type?: string;
  priority?: string;
  status?: UnifiedTestCaseStatus;
  stage?: UnifiedTestCaseStage;
  test_case_id?: string;
  module?: string;
  functional_module?: string;
  functional_domain?: string;
  preconditions?: string;  // 发送给后端时需要JSON字符串格式
  steps?: Array<{
    step_number: number;
    action: string;
    expected?: string;
  }>;
  expected_result?: string;
  remarks?: string;
  entity_order?: number;
}

export interface UnifiedTestCaseResponse extends UnifiedTestCaseBase {
  id: number;
  project_id: number;
  business_type: string;
  case_id: string;
  test_case_id: string; // 后端alias字段，与case_id相同
  stage: UnifiedTestCaseStage;
  module?: string;
  functional_module?: string;
  functional_domain?: string;
  preconditions: string | null;  // 后端返回JSON字符串格式
  steps: Array<{
    step_number: number;
    action: string;
    expected?: string;
  }> | null;
  expected_result: string | null;
  remarks?: string;
  generation_job_id?: string;
  entity_order?: number;
  created_at: string;
  updated_at: string;
}

export interface UnifiedTestCaseFilter extends PaginationConfig {
  project_id?: number;
  business_type?: string;
  status?: UnifiedTestCaseStatus;
  stage?: UnifiedTestCaseStage;
  priority?: PriorityLevel;
  keyword?: string;
  sort_by?: SortField;
  sort_order?: SortDirection;
  test_point_ids?: number[];
}

export interface UnifiedTestCaseListResponse extends PaginationResponse<UnifiedTestCaseResponse> {}

export interface UnifiedTestCaseStatistics {
  total_count: number;
  test_point_count: number;
  test_case_count: number;
  status_distribution: Record<string, number>;
  business_type_distribution: Record<string, number>;
  priority_distribution: Record<string, number>;
  recent_activity?: any[];
}

export interface UnifiedTestCaseBatchOperation {
  test_case_ids: number[];
  operation: 'delete' | 'update_status' | 'update_priority';
  status?: UnifiedTestCaseStatus;
  priority?: string;
}

export interface UnifiedTestCaseBatchResponse {
  success_count: number;
  failed_count: number;
  failed_items: Array<{
    case_id: number;
    error: string;
  }>;
}

export interface UnifiedTestCaseGenerationRequest {
  project_id: number;
  business_type: string;
  count?: number;
  complexity_level?: 'basic' | 'standard' | 'comprehensive';
  include_negative_cases?: boolean;
  additional_context?: Record<string, any>;
}

export interface UnifiedTestCaseGenerationResponse {
  generation_job_id: string;
  status: string;
  test_points_generated: number;
  test_cases_generated: number;
  test_cases?: UnifiedTestCaseResponse[]; // 修复：改为与当前数据库一致
  generation_time?: number;
  message: string;
}

// 表单相关类型
export interface UnifiedTestCaseFormData extends Omit<UnifiedTestCaseCreate, 'project_id'> {
  // 基础信息
  name: string;
  description: string;
  case_id: string;
  priority: 'low' | 'medium' | 'high';
  status: UnifiedTestCaseStatus;

  // 测试用例特有字段
  module?: string;
  functional_module?: string;
  functional_domain?: string;

  // 执行步骤
  preconditions?: string;  // 发送给后端时需要JSON字符串格式
  steps: Array<{
    step_number: number;
    action: string;
    expected: string;
  }>;
  expected_result?: string;
  remarks?: string;
}

// 视图模式
export enum ViewMode {
  TEST_POINTS = 'test_points',
  TEST_CASES = 'test_cases',
  UNIFIED = 'unified',
  HIERARCHY = 'hierarchy'
}

// 列表项扩展
export interface UnifiedTestCaseListItem extends UnifiedTestCaseResponse {
  selected?: boolean;
  loading?: boolean;
}

// 搜索和过滤
export interface SearchFilter {
  keyword?: string;
  business_type?: string;
  status?: UnifiedTestCaseStatus;
  priority?: string;
  stage?: UnifiedTestCaseStage;
  date_range?: [string, string];
}


// 表格列配置
export interface TableColumnConfig {
  key: string;
  title: string;
  dataIndex: string;
  width?: number;
  fixed?: 'left' | 'right';
  sorter?: boolean;
  filterable?: boolean;
  render?: (value: any, record: UnifiedTestCaseResponse) => React.ReactNode;
}

// 批量操作配置
export interface BatchAction {
  key: string;
  label: string;
  icon?: string;
  danger?: boolean;
  disabled?: boolean;
  onClick: (selectedIds: number[]) => Promise<void>;
}

// 导出配置
export interface ExportConfig {
  format: 'excel' | 'json' | 'csv';
  fields: string[];
  filename?: string;
}

// 统计图表数据
export interface ChartData {
  name: string;
  value: number;
  color?: string;
}

// 时间范围
export interface DateRange {
  start: string;
  end: string;
}

// 生成进度
export interface GenerationProgress {
  stage: 'test_point' | 'test_case' | 'completed';
  progress: number;
  message: string;
  estimated_time?: number;
}

// WebSocket消息
export interface UnifiedTestCaseWebSocketMessage {
  type: 'progress' | 'completed' | 'error' | 'status_update';
  data: any;
  timestamp: string;
}

// 表单验证规则
export interface FormRule {
  required?: boolean;
  message?: string;
  min?: number;
  max?: number;
  pattern?: RegExp;
  validator?: (rule: any, value: any) => Promise<void>;
}

// 统一错误响应格式
export interface UnifiedErrorResponse {
  error: string;
  code: string;
  details?: any;
  field?: string;
  timestamp?: string;
  trace_id?: string;
}

// API响应包装器
export interface ApiResponse<T = any> {
  success: boolean;
  data?: T;
  error?: UnifiedErrorResponse;
  message?: string;
  timestamp: string;
}

// 字段级错误
export interface FieldError {
  field: string;
  message: string;
  code?: string;
  value?: any;
}

// 验证错误响应
export interface ValidationErrorResponse extends UnifiedErrorResponse {
  field_errors: FieldError[];
}

File: src/models/unified_test_case.py
# -*- coding: utf-8 -*-
"""
Unified test case Pydantic models for API request/response serialization.

Represents both test points and test cases as the same entity at different stages:
- test_point: Initial stage with basic information (test_case_id + name + description)
- test_case: Complete stage with execution steps and details

They share the same id and test_case_id, differentiated only by the 'stage' field.
"""

from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, field_validator, ConfigDict
from enum import Enum
from src.validators.json_validators import JSONFieldValidator


class UnifiedTestCaseStatus(str, Enum):
    """Unified test case status for workflow management."""
    DRAFT = "draft"                    # Initial draft, needs review
    APPROVED = "approved"              # Approved and ready for test case generation
    COMPLETED = "completed"            # Test case fully completed with steps


class UnifiedTestCaseStage(str, Enum):
    """Test case stage to distinguish between test point and test case."""
    TEST_POINT = "test_point"          # Test point stage (basic info only)
    TEST_CASE = "test_case"            # Test case stage (with execution details)


# Base models
class UnifiedTestCaseBase(BaseModel):
    """Base unified test case model with common fields."""
    name: str = Field(..., min_length=1, max_length=200, description="测试用例名称")
    description: str = Field(..., min_length=1, max_length=2000, description="测试用例描述")
    priority: str = Field("medium", pattern="^(low|medium|high)$", description="优先级")
    status: UnifiedTestCaseStatus = Field(UnifiedTestCaseStatus.DRAFT, description="状态")


# Request models
class UnifiedTestCaseCreate(UnifiedTestCaseBase):
    """Unified test case creation model."""
    project_id: int = Field(..., gt=0, description="项目ID")
    business_type: str = Field(..., min_length=1, max_length=10, description="业务类型")
    test_case_id: str = Field(..., min_length=1, max_length=50, description="测试用例ID")

    # Test case specific fields (optional for test point stage)
    module: Optional[str] = Field(None, max_length=100, description="功能模块")
    functional_module: Optional[str] = Field(None, max_length=100, description="功能子模块")
    functional_domain: Optional[str] = Field(None, max_length=100, description="功能域")

    # Execution details (JSON format, optional for test point stage)
    preconditions: Optional[str] = Field(None, max_length=5000, description="前置条件")
    steps: Optional[List[Dict[str, Any]]] = Field(None, description="执行步骤")
    # Note: expected_result is removed - use expected field in steps instead
    remarks: Optional[str] = Field(None, max_length=2000, description="备注")

    # Metadata
    entity_order: Optional[float] = Field(None, description="排序顺序")

    
    @field_validator('steps')
    @classmethod
    def validate_steps(cls, v):
        return JSONFieldValidator.validate_steps(v)

    # expected_result field validation removed - field is removed from model


class UnifiedTestCaseUpdate(BaseModel):
    """Unified test case update model."""
    name: Optional[str] = Field(None, min_length=1, max_length=200, description="测试用例名称")
    description: Optional[str] = Field(None, max_length=2000, description="测试用例描述")
    business_type: Optional[str] = Field(None, min_length=1, max_length=20, description="业务类型")
    priority: Optional[str] = Field(None, pattern="^(low|medium|high)$", description="优先级")
    status: Optional[UnifiedTestCaseStatus] = Field(None, description="状态")

    # Test case specific fields
    module: Optional[str] = Field(None, max_length=100, description="功能模块")
    functional_module: Optional[str] = Field(None, max_length=100, description="功能子模块")
    functional_domain: Optional[str] = Field(None, max_length=100, description="功能域")

    # Execution details
    preconditions: Optional[str] = Field(None, max_length=5000, description="前置条件")
    steps: Optional[List[Dict[str, Any]]] = Field(None, description="执行步骤")
    # Note: expected_result is removed - use expected field in steps instead
    remarks: Optional[str] = Field(None, max_length=2000, description="备注")

    # Metadata
    entity_order: Optional[float] = Field(None, description="排序顺序")

    
    @field_validator('steps')
    @classmethod
    def validate_steps(cls, v):
        return JSONFieldValidator.validate_steps(v)

    # expected_result field validation removed - field is removed from model


# Response models
class UnifiedTestCaseResponse(UnifiedTestCaseBase):
    """Unified test case response model."""
    id: int = Field(..., description="测试用例ID")
    project_id: int = Field(..., description="项目ID")
    business_type: str = Field(..., description="业务类型")
    case_id: str = Field(..., alias="test_case_id", description="测试用例ID")
    test_case_id: str = Field(..., description="测试用例ID（与case_id相同，用于前端兼容）")

    # Stage information
    stage: UnifiedTestCaseStage = Field(..., description="当前阶段")

    # Test case specific fields (may be null for test point stage)
    module: Optional[str] = Field(None, description="功能模块")
    functional_module: Optional[str] = Field(None, description="功能子模块")
    functional_domain: Optional[str] = Field(None, description="功能域")

    # Execution details
    preconditions: Optional[str] = Field(None, max_length=5000, description="前置条件")
    steps: Optional[List[Dict[str, Any]]] = Field(None, description="执行步骤")
    # Note: expected_result is removed - use expected field in steps instead
    remarks: Optional[str] = Field(None, description="备注")

    # Metadata
    generation_job_id: Optional[str] = Field(None, description="生成任务ID")
    entity_order: Optional[float] = Field(None, description="排序顺序")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class UnifiedTestCaseListResponse(BaseModel):
    """Unified test case list response model."""
    items: List[UnifiedTestCaseResponse] = Field(..., description="测试用例列表")
    total: int = Field(..., description="总数量")
    page: int = Field(..., description="当前页码")
    size: int = Field(..., description="每页大小")
    pages: int = Field(..., description="总页数")


class UnifiedTestCaseFilter(BaseModel):
    """Unified test case filter parameters."""
    project_id: Optional[int] = Field(None, description="项目ID")
    business_type: Optional[str] = Field(None, description="业务类型")
    status: Optional[UnifiedTestCaseStatus] = Field(None, description="状态")
    stage: Optional[UnifiedTestCaseStage] = Field(None, description="阶段")
    priority: Optional[str] = Field(None, pattern="^(low|medium|high)$", description="优先级")
    keyword: Optional[str] = Field(None, description="关键词搜索")

    # Pagination
    page: int = Field(1, ge=1, description="页码")
    size: int = Field(20, ge=1, le=100, description="每页大小")

    # Sorting
    sort_by: str = Field("created_at", description="排序字段")
    sort_order: str = Field("desc", pattern="^(asc|desc)$", description="排序方向")


class UnifiedTestCaseStatistics(BaseModel):
    """Unified test case statistics model."""
    total_count: int = Field(..., description="总数量")
    test_point_count: int = Field(..., description="测试点阶段数量")
    test_case_count: int = Field(..., description="测试用例阶段数量")
    status_distribution: Dict[str, int] = Field(..., description="状态分布")
    business_type_distribution: Dict[str, int] = Field(..., description="业务类型分布")
    priority_distribution: Dict[str, int] = Field(..., description="优先级分布")


class UnifiedTestCaseBatchOperation(BaseModel):
    """Unified test case batch operation model."""
    test_case_ids: List[int] = Field(..., min_items=1, description="测试用例ID列表")
    operation: str = Field(..., pattern="^(delete|update_status|update_priority)$", description="操作类型")

    # Operation-specific parameters
    status: Optional[UnifiedTestCaseStatus] = Field(None, description="新状态（用于状态更新）")
    priority: Optional[str] = Field(None, pattern="^(low|medium|high)$", description="新优先级（用于优先级更新）")


class UnifiedTestCaseBatchResponse(BaseModel):
    """Unified test case batch operation response."""
    success_count: int = Field(..., description="成功操作数量")
    failed_count: int = Field(..., description="失败操作数量")
    failed_items: List[Dict[str, Any]] = Field(..., description="失败项目详情")


# Specialized generation models
class UnifiedTestCaseGenerationRequest(BaseModel):
    """Unified test case generation request."""
    project_id: int = Field(..., gt=0, description="项目ID")
    business_type: str = Field(..., description="业务类型")

    # Generation mode: test_points_only or test_cases_only
    generation_mode: str = Field(..., pattern="^(test_points_only|test_cases_only)$", description="生成模式")

    # Test point IDs (required for test_cases_only mode)
    test_point_ids: Optional[List[int]] = Field(None, description="测试点ID列表（仅用于test_cases_only模式）")

    # Additional context as simple string
    additional_context: Optional[str] = Field(None, max_length=2000, description="额外上下文")


class UnifiedTestCaseGenerationResponse(BaseModel):
    """Unified test case generation response."""
    generation_job_id: str = Field(..., description="生成任务ID")
    status: str = Field(..., description="任务状态")
    test_points_generated: int = Field(..., description="生成的测试点数量")
    test_cases_generated: int = Field(..., description="生成的测试用例数量")
    unified_test_cases: Optional[List[UnifiedTestCaseResponse]] = Field(None, description="生成的测试用例列表")
    generation_time: Optional[float] = Field(None, description="生成耗时（秒）")
    message: str = Field(..., description="状态消息")


class UnifiedTestCaseDeleteResponse(BaseModel):
    """Response model for delete operations."""
    message: str